Failures when the unit tests are run with race flag

Changed the test-unit target in Makefile to run with race flag: "$(GO) test -race ./..."

There are quite a few data race failures!! Logs are below.


I1126 06:25:16.101098   63379 udp.go:80] Collecting process from 127.0.0.1:52940 has stopped.
I1126 06:25:18.099195   63379 process.go:262] Template with id 256, and obsDomainID 1 is expired.

WARNING: DATA RACE
Write at 0x00c000115c20 by goroutine 63:
  runtime.mapdelete()
      /usr/local/go/src/runtime/map.go:685 +0x0
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).deleteTemplate()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process.go:282 +0x138
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).addTemplate.func1()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process.go:263 +0x214

Previous read at 0x00c000115c20 by goroutine 57:
  runtime.mapaccess1()
      /usr/local/go/src/runtime/map.go:394 +0x0
  github.com/vmware/go-ipfix/pkg/collector.TestUDPCollectingProcess_TemplateExpire()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process_test.go:367 +0x2d0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 63 (running) created at:
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).addTemplate()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process.go:257 +0x44e
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).decodeTemplateSet()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process.go:207 +0xc11
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).decodePacket()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/process.go:130 +0x404
  github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).handleUDPClient.func1()
      /Users/stati/work/antrea_all/go-ipfix/pkg/collector/udp.go:88 +0x2c7

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:111 +0x356

--- FAIL: TestUDPCollectingProcess_TemplateExpire (15.01s)
    testing.go:1038: race detected during execution of test
FAIL
coverage: 87.6% of statements
FAIL	github.com/vmware/go-ipfix/pkg/collector	42.291s
ok  	github.com/vmware/go-ipfix/pkg/entities	1.781s	coverage: 69.7% of statements
?   	github.com/vmware/go-ipfix/pkg/entities/testing	[no test files]

WARNING: DATA RACE
Read at 0x00c000113168 by goroutine 12:
  github.com/vmware/go-ipfix/pkg/intermediate.(*AggregationProcess).Stop()
      /Users/stati/work/antrea_all/go-ipfix/pkg/intermediate/aggregate.go:71 +0x8a
  github.com/vmware/go-ipfix/pkg/intermediate.TestAggregationProcess.func1()
      /Users/stati/work/antrea_all/go-ipfix/pkg/intermediate/aggregate_test.go:207 +0x97

Previous write at 0x00c000113168 by goroutine 11:
  github.com/vmware/go-ipfix/pkg/intermediate.(*AggregationProcess).Start()
      /Users/stati/work/antrea_all/go-ipfix/pkg/intermediate/aggregate.go:65 +0x2b1
  github.com/vmware/go-ipfix/pkg/intermediate.TestAggregationProcess()
      /Users/stati/work/antrea_all/go-ipfix/pkg/intermediate/aggregate_test.go:211 +0x11e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 12 (running) created at:
  github.com/vmware/go-ipfix/pkg/intermediate.TestAggregationProcess()
      /Users/stati/work/antrea_all/go-ipfix/pkg/intermediate/aggregate_test.go:203 +0x10d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 11 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:105 +0x356
--- FAIL: TestAggregationProcess (0.00s)
    testing.go:1038: race detected during execution of test
FAIL
coverage: 80.3% of statements
FAIL	github.com/vmware/go-ipfix/pkg/intermediate	0.468s
ok  	github.com/vmware/go-ipfix/pkg/producer	2.036s	coverage: 72.9% of statements
?   	github.com/vmware/go-ipfix/pkg/producer/protobuf	[no test files]
ok  	github.com/vmware/go-ipfix/pkg/registry	1.537s	coverage: 99.4% of statements
?   	github.com/vmware/go-ipfix/pkg/util	[no test files]


pkg/intermediate, pkg/collector, and pkg/exporter have failures. 

We should fix these with priority and turn on the race flag when unit tests are run.

I can take a crack at pkg/exporter. @zyiou could you look into pkg/collector failures?

--- a/Makefile
+++ b/Makefile
@@ -12,10 +12,10 @@ codegen:
 	mkdir -p ./.coverage
 
 test-unit: .coverage
-	$(GO) test ./... -covermode=atomic -coverprofile=.coverage/coverage_unit.txt
+	$(GO) test -race ./... -covermode=atomic -coverprofile=.coverage/coverage_unit.txt
 
 test-integration: .coverage
-	$(GO) test ./pkg/test/... -tags=integration -covermode=atomic -coverprofile=.coverage/coverage_integration.txt -coverpkg github.com/vmware/go-ipfix/pkg/collector,github.com/vmware/go-ipfix/pkg/exporter,github.com/vmware/go-ipfix/pkg/intermediate
+	$(GO) test -race ./pkg/test/... -tags=integration -covermode=atomic -coverprofile=.coverage/coverage_integration.txt -coverpkg github.com/vmware/go-ipfix/pkg/collector,github.com/vmware/go-ipfix/pkg/exporter,github.com/vmware/go-ipfix/pkg/intermediate
 
 golangci:
 	@curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.21.0

--- a/go.mod
+++ b/go.mod
@@ -7,6 +7,7 @@ require (
 	github.com/spf13/cobra v0.0.5
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.5.1
+	k8s.io/apimachinery v0.18.4
 	k8s.io/component-base v0.18.4
 	k8s.io/klog v1.0.0
 )

--- a/pkg/collector/process.go
+++ b/pkg/collector/process.go
@@ -33,8 +33,8 @@ import (
 type CollectingProcess struct {
 	// for each obsDomainID, there is a map of templates
 	templatesMap map[uint32]map[uint16][]*entities.InfoElement
-	// templatesLock allows multiple readers or one writer at the same time
-	templatesLock sync.RWMutex
+	// mutex allows multiple readers or one writer at the same time
+	mutex sync.RWMutex
 	// template lifetime
 	templateTTL uint32
 	// server information
@@ -47,8 +47,6 @@ type CollectingProcess struct {
 	messageChan chan *entities.Message
 	// maps each client to its client handler (required channels)
 	clients map[string]*clientHandler
-	// clientsLock allows multiple readers or one writer to access clients map at the same time
-	clientsLock sync.RWMutex
 }
 
 type clientHandler struct {
@@ -59,7 +57,7 @@ type clientHandler struct {
 func InitCollectingProcess(address net.Addr, maxBufferSize uint16, templateTTL uint32) (*CollectingProcess, error) {
 	collectProc := &CollectingProcess{
 		templatesMap:  make(map[uint32]map[uint16][]*entities.InfoElement),
-		templatesLock: sync.RWMutex{},
+		mutex:         sync.RWMutex{},
 		templateTTL:   templateTTL,
 		address:       address,
 		maxBufferSize: maxBufferSize,
@@ -80,13 +78,16 @@ func (cp *CollectingProcess) Start() {
 
 func (cp *CollectingProcess) Stop() {
 	cp.stopChan <- true
-	if cp.messageChan != nil {
-		close(cp.messageChan)
-	}
 }
 
 func (cp *CollectingProcess) GetMsgChan() chan *entities.Message {
 	return cp.messageChan
+}
+
+func (cp *CollectingProcess) CloseMsgChan() {
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
+	close(cp.messageChan)
 }
 
 func (cp *CollectingProcess) createClient() *clientHandler {
@@ -97,21 +98,29 @@ func (cp *CollectingProcess) createClient() *clientHandler {
 }
 
 func (cp *CollectingProcess) addClient(address string, client *clientHandler) {
-	cp.clientsLock.Lock()
-	defer cp.clientsLock.Unlock()
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
 	cp.clients[address] = client
 }
 
 func (cp *CollectingProcess) deleteClient(name string) {
-	cp.clientsLock.Lock()
-	defer cp.clientsLock.Unlock()
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
 	delete(cp.clients, name)
 }
 
 func (cp *CollectingProcess) getClientCount() int {
-	cp.clientsLock.RLock()
-	defer cp.clientsLock.RUnlock()
+	cp.mutex.RLock()
+	defer cp.mutex.RUnlock()
 	return len(cp.clients)
+}
+
+func (cp *CollectingProcess) closeAllClients() {
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
+	for _, client := range cp.clients {
+		client.errChan <- true
+	}
 }
 
 func (cp *CollectingProcess) decodePacket(packetBuffer *bytes.Buffer, exportAddress string) (*entities.Message, error) {
@@ -243,7 +252,8 @@ func (cp *CollectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID
 }
 
 func (cp *CollectingProcess) addTemplate(obsDomainID uint32, templateID uint16, elementsWithValue []*entities.InfoElementWithValue) {
-	cp.templatesLock.Lock()
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
 	if _, exists := cp.templatesMap[obsDomainID]; !exists {
 		cp.templatesMap[obsDomainID] = make(map[uint16][]*entities.InfoElement)
 	}
@@ -252,7 +262,6 @@ func (cp *CollectingProcess) addTemplate(obsDomainID uint32, templateID uint16, 
 		elements = append(elements, elementWithValue.Element)
 	}
 	cp.templatesMap[obsDomainID][templateID] = elements
-	cp.templatesLock.Unlock()
 	// template lifetime management
 	if cp.address.Network() == "tcp" {
 		return
@@ -275,8 +284,8 @@ func (cp *CollectingProcess) addTemplate(obsDomainID uint32, templateID uint16, 
 }
 
 func (cp *CollectingProcess) getTemplate(obsDomainID uint32, templateID uint16) ([]*entities.InfoElement, error) {
-	cp.templatesLock.RLock()
-	defer cp.templatesLock.RUnlock()
+	cp.mutex.RLock()
+	defer cp.mutex.RUnlock()
 	if elements, exists := cp.templatesMap[obsDomainID][templateID]; exists {
 		return elements, nil
 	} else {
@@ -285,8 +294,8 @@ func (cp *CollectingProcess) getTemplate(obsDomainID uint32, templateID uint16) 
 }
 
 func (cp *CollectingProcess) deleteTemplate(obsDomainID uint32, templateID uint16) {
-	cp.templatesLock.Lock()
-	defer cp.templatesLock.Unlock()
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
 	delete(cp.templatesMap[obsDomainID], templateID)
 }
 

--- a/pkg/collector/process_test.go
+++ b/pkg/collector/process_test.go
@@ -22,6 +22,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/vmware/go-ipfix/pkg/entities"
 	"github.com/vmware/go-ipfix/pkg/registry"
@@ -48,28 +49,22 @@ func TestTCPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 	if err != nil {
 		t.Fatalf("TCP Collecting Process does not start correctly: %v", err)
 	}
-	messageCount := 0
+	go cp.Start()
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
+
 	go func() {
-		time.Sleep(2 * time.Second)
 		conn, err := net.Dial(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("Cannot establish connection to %s", address.String())
 		}
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
-		time.Sleep(time.Second)
-		msgChan := cp.GetMsgChan()
-		for range msgChan {
-			messageCount++
-		}
 	}()
-	go func() {
-		time.Sleep(4 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
-	assert.NotNil(t, cp.templatesMap[1], "TCP Collecting Process should receive and store the received template.")
-	assert.Equal(t, 1, messageCount, "Messages should be stored correctly.")
+	<-cp.GetMsgChan()
+	cp.Stop()
+	template, _ := cp.getTemplate(1, 256)
+	assert.NotNil(t, template, "TCP Collecting Process should receive and store the received template.")
 }
 
 func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
@@ -81,9 +76,11 @@ func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 	if err != nil {
 		t.Fatalf("UDP Collecting Process does not start correctly: %v", err)
 	}
-	messageCount := 0
+	go cp.Start()
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
+
 	go func() {
-		time.Sleep(2 * time.Second)
 		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
@@ -94,19 +91,12 @@ func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 		}
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
-		time.Sleep(time.Second)
-		msgChan := cp.GetMsgChan()
-		for range msgChan {
-			messageCount++
-		}
 	}()
-	go func() {
-		time.Sleep(4 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
-	assert.NotNil(t, cp.templatesMap[1], "UDP Collecting Process should receive and store the received template.")
-	assert.Equal(t, 1, messageCount, "Messages should be stored correctly.")
+	<-cp.GetMsgChan()
+	cp.Stop()
+	template, _ := cp.getTemplate(1, 256)
+	assert.NotNil(t, template, "UDP Collecting Process should receive and store the received template.")
+
 }
 
 func TestTCPCollectingProcess_ReceiveDataRecord(t *testing.T) {
@@ -114,33 +104,28 @@ func TestTCPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	messageCount := 0
 	cp, err := InitCollectingProcess(address, 1024, 0)
 	// Add the templates before sending data record
 	cp.addTemplate(uint32(1), uint16(256), elementsWithValue)
 	if err != nil {
 		t.Fatalf("TCP Collecting Process does not start correctly: %v", err)
 	}
+
+	go cp.Start()
+
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
+
 	go func() {
-		time.Sleep(time.Second)
 		conn, err := net.Dial(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("Cannot establish connection to %s", address.String())
 		}
 		defer conn.Close()
 		conn.Write(validDataPacket)
-		time.Sleep(time.Second)
-		msgChan := cp.GetMsgChan()
-		for range msgChan {
-			messageCount++
-		}
 	}()
-	go func() {
-		time.Sleep(4 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
-	assert.Equal(t, 1, messageCount, "TCP Collecting Process should receive and store the received data record.")
+	<-cp.GetMsgChan()
+	cp.Stop()
 }
 
 func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
@@ -148,15 +133,18 @@ func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	messageCount := 0
 	cp, err := InitCollectingProcess(address, 1024, 0)
 	// Add the templates before sending data record
 	cp.addTemplate(uint32(1), uint16(256), elementsWithValue)
 	if err != nil {
 		t.Fatalf("UDP Collecting Process does not start correctly: %v", err)
 	}
+
+	go cp.Start()
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
+
 	go func() {
-		time.Sleep(time.Second)
 		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
@@ -167,16 +155,9 @@ func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 		}
 		defer conn.Close()
 		conn.Write(validDataPacket)
-		for range cp.GetMsgChan() {
-			messageCount++
-		}
 	}()
-	go func() {
-		time.Sleep(5 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
-	assert.Equal(t, 1, messageCount, "UDP Collecting Process should receive and store the received data record.")
+	<-cp.GetMsgChan()
+	cp.Stop()
 }
 
 func TestTCPCollectingProcess_ConcurrentClient(t *testing.T) {
@@ -186,20 +167,22 @@ func TestTCPCollectingProcess_ConcurrentClient(t *testing.T) {
 	}
 	cp, _ := InitCollectingProcess(address, 1024, 0)
 	go func() {
-		time.Sleep(time.Second)
+		// wait until collector is ready
+		waitForCollectorReady(t, address)
 		_, err := net.Dial(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("Cannot establish connection to %s", address.String())
 		}
 	}()
 	go func() {
-		time.Sleep(time.Second)
+		// wait until collector is ready
+		waitForCollectorReady(t, address)
 		_, err := net.Dial(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("Cannot establish connection to %s", address.String())
 		}
-		time.Sleep(2 * time.Second)
-		assert.Equal(t, 2, cp.getClientCount(), "There should be two tcp clients.")
+		time.Sleep(time.Millisecond)
+		assert.Equal(t, 4, cp.getClientCount(), "There should be 4 tcp clients.")
 		cp.Stop()
 	}()
 	cp.Start()
@@ -211,8 +194,10 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 		t.Error(err)
 	}
 	cp, _ := InitCollectingProcess(address, 1024, 0)
+	go cp.Start()
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
 	go func() {
-		time.Sleep(time.Second)
 		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
@@ -223,12 +208,8 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 		}
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
-		// remove the messages from the message channel
-		for range cp.GetMsgChan() {
-		}
 	}()
 	go func() {
-		time.Sleep(time.Second)
 		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
@@ -239,20 +220,19 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 		}
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
-		time.Sleep(time.Second)
-		assert.Equal(t, 2, len(cp.clients), "There should be two tcp clients.")
+		time.Sleep(time.Millisecond)
+		assert.Equal(t, 2, cp.getClientCount(), "There should be two tcp clients.")
 	}()
-	go func() {
-		time.Sleep(6 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
+	// there should be two messages received
+	<-cp.GetMsgChan()
+	<-cp.GetMsgChan()
+	cp.Stop()
 }
 
 func TestCollectingProcess_DecodeTemplateRecord(t *testing.T) {
 	cp := CollectingProcess{}
 	cp.templatesMap = make(map[uint32]map[uint16][]*entities.InfoElement)
-	cp.templatesLock = sync.RWMutex{}
+	cp.mutex = sync.RWMutex{}
 	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4736")
 	if err != nil {
 		t.Error(err)
@@ -293,7 +273,7 @@ func TestCollectingProcess_DecodeTemplateRecord(t *testing.T) {
 func TestCollectingProcess_DecodeDataRecord(t *testing.T) {
 	cp := CollectingProcess{}
 	cp.templatesMap = make(map[uint32]map[uint16][]*entities.InfoElement)
-	cp.templatesLock = sync.RWMutex{}
+	cp.mutex = sync.RWMutex{}
 	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4737")
 	if err != nil {
 		t.Error(err)
@@ -331,12 +311,14 @@ func TestUDPCollectingProcess_TemplateExpire(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	cp, err := InitCollectingProcess(address, 1024, 5)
+	cp, err := InitCollectingProcess(address, 1024, 1)
 	if err != nil {
 		t.Fatalf("UDP Collecting Process does not start correctly: %v", err)
 	}
+	go cp.Start()
+	// wait until collector is ready
+	waitForCollectorReady(t, address)
 	go func() {
-		time.Sleep(2 * time.Second)
 		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
@@ -350,17 +332,26 @@ func TestUDPCollectingProcess_TemplateExpire(t *testing.T) {
 		if err != nil {
 			t.Errorf("Error in sending data to collector: %v", err)
 		}
-		go func() { // remove the message from the message channel
-			for range cp.GetMsgChan() {
-			}
-		}()
 	}()
-	go func() {
-		time.Sleep(5 * time.Second)
-		cp.Stop()
-	}()
-	cp.Start()
-	assert.NotNil(t, cp.templatesMap[1][256], "Template should be stored in the template map.")
-	time.Sleep(10 * time.Second)
-	assert.Nil(t, cp.templatesMap[1][256], "Template should be deleted after 5 seconds.")
+	<-cp.GetMsgChan()
+	cp.Stop()
+	template, err := cp.getTemplate(1, 256)
+	assert.NotNil(t, template, "Template should be stored in the template map.")
+	assert.Nil(t, err, "Template should be stored in the template map.")
+	time.Sleep(2 * time.Second)
+	template, err = cp.getTemplate(1, 256)
+	assert.Nil(t, template, "Template should be deleted after 5 seconds.")
+	assert.NotNil(t, err, "Template should be deleted after 5 seconds.")
+}
+
+func waitForCollectorReady(t *testing.T, address net.Addr) {
+	checkConn := func() (bool, error) {
+		if _, err := net.Dial(address.Network(), address.String()); err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	if err := wait.Poll(100*time.Millisecond, 500*time.Millisecond, checkConn); err != nil {
+		t.Errorf("Cannot establish connection to %s", address.String())
+	}
 }

--- a/pkg/collector/tcp.go
+++ b/pkg/collector/tcp.go
@@ -30,14 +30,10 @@ func (cp *CollectingProcess) startTCPServer() {
 			go cp.handleTCPClient(conn, &wg)
 		}
 	}()
-	if <-cp.stopChan {
-		// close all connections
-		for _, client := range cp.clients {
-			client.errChan <- true
-		}
-		wg.Wait()
-		return
-	}
+	<-cp.stopChan
+	// close all connections
+	cp.closeAllClients()
+	wg.Wait()
 }
 
 func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) {
@@ -78,8 +74,6 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) 
 			}
 		}
 	}()
-	if <-client.errChan {
-		cp.deleteClient(address)
-		return
-	}
+	<-client.errChan
+	cp.deleteClient(address)
 }

--- a/pkg/collector/udp.go
+++ b/pkg/collector/udp.go
@@ -55,15 +55,10 @@ func (cp *CollectingProcess) startUDPServer() {
 			cp.clients[address.String()].packetChan <- bytes.NewBuffer(buff[0:size])
 		}
 	}()
-	select {
-	case <-cp.stopChan:
-		// stop all the workers before closing collector
-		for _, client := range cp.clients {
-			client.errChan <- true
-		}
-		wg.Wait()
-		return
-	}
+	<-cp.stopChan
+	// stop all the workers before closing collector
+	cp.closeAllClients()
+	wg.Wait()
 }
 
 func (cp *CollectingProcess) handleUDPClient(address net.Addr, wg *sync.WaitGroup) {
@@ -81,7 +76,7 @@ func (cp *CollectingProcess) handleUDPClient(address net.Addr, wg *sync.WaitGrou
 					return
 				case <-ticker.C: // set timeout for udp connection
 					klog.Errorf("UDP connection from %s timed out.", address.String())
-					cp.deleteClient(address.Network())
+					cp.deleteClient(address.String())
 					return
 				case packet := <-client.packetChan:
 					// get the message here

--- a/pkg/exporter/process_test.go
+++ b/pkg/exporter/process_test.go
@@ -263,7 +263,7 @@ func TestExportingProcess_SendingDataRecordToLocalUDPServer(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Got error when resolving UDP address: %v", err)
 	}
-	conn, err := net.ListenUDP("udp", udpAddr)
+	conn, _ := net.ListenUDP("udp", udpAddr)
 	t.Log("Created local server on random available port for testing")
 
 	buffCh := make(chan []byte)

--- a/pkg/intermediate/aggregate.go
+++ b/pkg/intermediate/aggregate.go
@@ -15,8 +15,8 @@ import (
 type AggregationProcess struct {
 	// flowKeyRecordMap maps each connection (5-tuple) with its records
 	flowKeyRecordMap map[FlowKey][]entities.Record
-	// flowKeyRecordLock allows multiple readers or one writer at the same time
-	flowKeyRecordLock sync.RWMutex
+	// mutex allows multiple readers or one writer at the same time
+	mutex sync.RWMutex
 	// messageChan is the channel to receive the message
 	messageChan chan *entities.Message
 	// workerNum is the number of workers to process the messages
@@ -59,18 +59,22 @@ func InitAggregationProcess(messageChan chan *entities.Message, workerNum int, c
 }
 
 func (a *AggregationProcess) Start() {
+	a.mutex.Lock()
 	for i := 0; i < a.workerNum; i++ {
 		w := createWorker(i, a.messageChan, a.AggregateMsgByFlowKey)
 		w.start()
 		a.workerList = append(a.workerList, w)
 	}
+	a.mutex.Unlock()
 	<-a.stopChan
 }
 
 func (a *AggregationProcess) Stop() {
+	a.mutex.Lock()
 	for _, worker := range a.workerList {
 		worker.stop()
 	}
+	a.mutex.Unlock()
 	a.stopChan <- true
 }
 
@@ -96,8 +100,8 @@ func (a *AggregationProcess) AggregateMsgByFlowKey(message *entities.Message) er
 
 // ForAllRecordsDo takes in callback function to process the operations to flowkey->records pairs in the map
 func (a *AggregationProcess) ForAllRecordsDo(callback FlowKeyRecordMapCallBack) error {
-	a.flowKeyRecordLock.RLock()
-	defer a.flowKeyRecordLock.RUnlock()
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
 	for k, v := range a.flowKeyRecordMap {
 		err := callback(k, v)
 		if err != nil {
@@ -109,15 +113,15 @@ func (a *AggregationProcess) ForAllRecordsDo(callback FlowKeyRecordMapCallBack) 
 }
 
 func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
-	a.flowKeyRecordLock.Lock()
-	defer a.flowKeyRecordLock.Unlock()
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
 	delete(a.flowKeyRecordMap, flowKey)
 }
 
 // correlateRecords fills records info by correlating incoming and current records
 func (a *AggregationProcess) correlateRecords(flowKey FlowKey, record entities.Record) {
-	a.flowKeyRecordLock.Lock()
-	defer a.flowKeyRecordLock.Unlock()
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
 	existingRecords := a.flowKeyRecordMap[flowKey]
 	// only fill the information for record from source node
 	if isRecordFromSrc(record) {

--- a/pkg/test/collector_intermediate_test.go
+++ b/pkg/test/collector_intermediate_test.go
@@ -1,16 +1,33 @@
+// Copyright 2020 VMware, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build integration
+
 package test
 
 import (
+	"fmt"
 	"net"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
-
 	"github.com/vmware/go-ipfix/pkg/collector"
 	"github.com/vmware/go-ipfix/pkg/entities"
 	"github.com/vmware/go-ipfix/pkg/intermediate"
 	"github.com/vmware/go-ipfix/pkg/registry"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 var templatePacket = []byte{0, 10, 0, 76, 95, 154, 84, 121, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 60, 1, 0, 0, 9, 0, 8, 0, 4, 0, 12, 0, 4, 0, 7, 0, 2, 0, 11, 0, 2, 0, 4, 0, 1, 128, 101, 255, 255, 0, 0, 220, 186, 128, 103, 255, 255, 0, 0, 220, 186, 128, 106, 0, 4, 0, 0, 220, 186, 128, 107, 0, 2, 0, 0, 220, 186}
@@ -32,30 +49,23 @@ func TestCollectorToIntermediate(t *testing.T) {
 	// Initialize aggregation process and collecting process
 	cp, _ := collector.InitCollectingProcess(address, 1024, 0)
 	ap, _ := intermediate.InitAggregationProcess(cp.GetMsgChan(), 2, fields)
-
+	go cp.Start()
+	waitForCollectorReady(t, address)
 	go func() {
-		time.Sleep(time.Second)
 		conn, err := net.DialUDP("udp", nil, address)
 		if err != nil {
 			t.Errorf("UDP Collecting Process does not start correctly.")
 		}
 		defer conn.Close()
 		conn.Write(templatePacket)
-		time.Sleep(time.Second)
 		conn.Write(dataPacket1)
-		time.Sleep(time.Second)
 		conn.Write(dataPacket2)
 	}()
-	go func() {
-		ap.Start()
-	}()
-	go func() {
-		time.Sleep(4 * time.Second)
-		cp.Stop()
-		ap.Stop()
-	}()
-	cp.Start()
-	ap.ForAllRecordsDo(copyFlowKeyRecordMap)
+	go ap.Start()
+	waitForAggrationFinished(t, ap)
+	cp.Stop()
+	ap.Stop()
+
 	assert.Equal(t, 1, len(flowKeyRecordMap), "Aggregation process should store the data record to map with corresponding flow key.")
 	flowKey := intermediate.FlowKey{SourceAddress: "1.2.3.4", DestinationAddress: "5.6.7.8", Protocol: 6, SourcePort: 1234, DestinationPort: 5678}
 	assert.NotNil(t, flowKeyRecordMap[flowKey])
@@ -73,4 +83,30 @@ func TestCollectorToIntermediate(t *testing.T) {
 func copyFlowKeyRecordMap(key intermediate.FlowKey, records []entities.Record) error {
 	flowKeyRecordMap[key] = records
 	return nil
+}
+
+func waitForCollectorReady(t *testing.T, address net.Addr) {
+	checkConn := func() (bool, error) {
+		if _, err := net.Dial(address.Network(), address.String()); err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	if err := wait.Poll(100 * time.Millisecond, 500 * time.Millisecond, checkConn); err != nil {
+		t.Errorf("Cannot establish connection to %s", address.String())
+	}
+}
+
+func waitForAggrationFinished(t *testing.T, ap *intermediate.AggregationProcess) {
+	checkConn := func() (bool, error) {
+		ap.ForAllRecordsDo(copyFlowKeyRecordMap)
+		if len(flowKeyRecordMap) > 0 {
+			return true, nil
+		} else {
+			return false, fmt.Errorf("aggregation process does not process and store data correctly")
+		}
+	}
+	if err := wait.Poll(100 * time.Millisecond, 500 * time.Millisecond, checkConn); err != nil {
+		t.Error(err)
+	}
 }

--- a/pkg/test/exporter_collector_test.go
+++ b/pkg/test/exporter_collector_test.go
@@ -19,7 +19,6 @@ package test
 import (
 	"net"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/assert"
 	"k8s.io/klog"
@@ -29,6 +28,11 @@ import (
 	"github.com/vmware/go-ipfix/pkg/exporter"
 	"github.com/vmware/go-ipfix/pkg/registry"
 )
+
+func init() {
+	// Load the global registry
+	registry.LoadRegistry()
+}
 
 func TestSingleRecordUDPTransport(t *testing.T) {
 	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
@@ -63,14 +67,13 @@ func TestMultipleRecordTCPTransport(t *testing.T) {
 }
 
 func testExporterToCollector(address net.Addr, isMultipleRecord bool, t *testing.T) {
-	// Load the global registry
-	registry.LoadRegistry()
 	// Initialize collecting process
 	messages := make([]*entities.Message, 0)
 	cp, _ := collector.InitCollectingProcess(address, 1024, 0)
-
+	// Start collecting process
+	go cp.Start()
 	go func() { // Start exporting process in go routine
-		time.Sleep(2 * time.Second) // wait for collector to be ready
+		waitForCollectorReady(t, address)
 		export, err := exporter.InitExportingProcess(address, 1, 0)
 		if err != nil {
 			klog.Fatalf("Got error when connecting to %s", address.String())
@@ -121,7 +124,6 @@ func testExporterToCollector(address net.Addr, isMultipleRecord bool, t *testing
 		if err != nil {
 			klog.Fatalf("Got error when sending record: %v", err)
 		}
-		time.Sleep(time.Second)
 		// Create data set with 1 data record using the same template above
 		dataSet := entities.NewSet(entities.Data, templateID, false)
 		elements = make([]*entities.InfoElementWithValue, 0)
@@ -208,16 +210,15 @@ func testExporterToCollector(address net.Addr, isMultipleRecord bool, t *testing
 			klog.Fatalf("Got error when sending record: %v", err)
 		}
 		export.CloseConnToCollector() // Close exporting process
-		time.Sleep(2 * time.Second)
-		cp.Stop() // Close collecting process
 	}()
-	go func() {
-		for message := range cp.GetMsgChan() {
-			messages = append(messages, message)
+
+	for message := range cp.GetMsgChan() {
+		messages = append(messages, message)
+		if len(messages) == 2 {
+			cp.CloseMsgChan()
 		}
-	}()
-	// Start collecting process
-	cp.Start()
+	}
+	cp.Stop() // Close collecting process
 	templateMsg := messages[0]
 
 	assert.Equal(t, uint16(10), templateMsg.GetVersion(), "Version of flow record (template) should be 10.")

